src/cdogs/pickup.c

```diff
@@ -356,6 +356,16 @@ static bool TryPickupGun(
 			: StrWeaponClass(AmmoGetById(&gAmmo, pe->u.Ammo.Id)->DefaultGun);
 	const int actorsGunIdx = ActorFindGun(a, wc);
 
+	// Don't pick up guns in the melee slot or melee weapons in a gun slot
+	if (a->gunIndex == MELEE_SLOT && wc->Type != GUNTYPE_MELEE)
```

Comment:
there's a third type of weapon - grenades, will need to avoid switching to grenades, since you can never have that slot selected, you can only throw grenades using the grenade button

Reply:
I need clarification so I don't get this wrong. The checks I've added take place before the original logic for weapon pickup. The grenade pickup should be unchanged from the original logic. Is it currently possible to pick up anything other than grenades in the grenade slot? If not, I'm not sure what I should change with what I've added.

Reply:
no, only grenades/explosives go in the grenade slot
I highlighted this line because it looks like it would allow picking up grenades into the melee slot if that's what the player is holding
also to correct what I said earlier, it is actually possible to switch into grenades, if the player doesn't have a grenade key bound (I think this is only possible if players use old key bindings, newer versions of the game come with a default grenade key bind and you can't unbind keys to nothing in the options menu, unless you edit the players.cnf file manually)

Reply:
so this would be a testing matrix:

| Holding / Picking up         | Gun           | Melee | Grenade  |
| ------------- |:-------------:| -----:| ---:|
| Gun      | Pickup into empty slot or swap with current gun, switch to new gun | Add or swap melee, switch to melee | Add or swap grenade, switch to grenade only if able (no grenade key bind) |
| Melee     | Pickup into empty slot or swap with 0th gun, switch to new gun      | Add or swap melee | Add or swap grenade, switch to grenade only if able (no grenade key bind) |
| Grenade (i.e. has no grenade key bind)  | Pickup into empty slot or swap with 0th gun, switch to new gun     | Add or swap melee, switch to melee | Add or swap grenade |

Reply:
or another way of looking at this is to maintain the following:

- Picking up weapons always results in switching to that weapon (except for grenades if grenades can't be switched to)
- If there is an empty slot, always pickup into that slot (only applicable for guns because there are 2 gun slots)
- If there are no empty slots, drop the currently selected weapon and pickup the new weapon into the same slot

Reply:
For testing you can use `custom/techdemo/Spawner Test` which contains a bunch of weapon pickups

Reply:
I see what you mean now after some testing, will work on this.